hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/TestCloseContainer.java

```diff
@@ -159,11 +182,98 @@ public void testCloseClosedContainer()
+    ContainerProtos.ContainerChecksumInfo expectedChecksumInfo1 = ContainerMerkleTreeTestUtils.readChecksumFile(
+        container1.getContainerData());
+
+    // Checksum file exists after container close and matches the expected container
+    // merkle tree for all the datanodes
+    for (HddsDatanodeService hddsDatanode : hddsDatanodes) {
+      GenericTestUtils.waitFor(() -> waitForContainerChecksumToExist(hddsDatanode, containerInfo1), 100, 5000);
+      container1 = (KeyValueContainer) ozoneContainer.getController().getContainer(containerInfo1.getContainerID());
+      ContainerProtos.ContainerChecksumInfo containerChecksumInfo = ContainerMerkleTreeTestUtils.readChecksumFile(
+              container1.getContainerData());
+      ContainerMerkleTreeTestUtils.assertTreesSortedAndMatch(expectedChecksumInfo1.getContainerMerkleTree(),
+          containerChecksumInfo.getContainerMerkleTree());
+    }
+
+    // Create 2nd container and check the checksum doesn't match with 1st container
+    TestDataUtil.createKey(bucket, "key2", ReplicationFactor.THREE,
+        ReplicationType.RATIS, "this is the different content");
+    ContainerInfo containerInfo2 = scm.getContainerManager().getContainers().get(1);
+    for (HddsDatanodeService hddsDatanode : hddsDatanodes) {
+      assertFalse(containerChecksumFileExists(hddsDatanode, containerInfo2));
+    }
+
+    // Close container.
+    OzoneTestUtils.closeContainer(scm, containerInfo2);
+    // Build expected merkle tree for one datanode.
+    KeyValueContainer container2 = (KeyValueContainer) ozoneContainer.getController()
+        .getContainer(containerInfo2.getContainerID());
+    ContainerProtos.ContainerChecksumInfo expectedChecksumInfo2 = ContainerMerkleTreeTestUtils.readChecksumFile(
+        container2.getContainerData());
+
+    // Container merkle tree for different container should not match.
+    assertNotEquals(expectedChecksumInfo1.getContainerID(), expectedChecksumInfo2.getContainerID());
+    assertNotEquals(expectedChecksumInfo1.getContainerMerkleTree().getDataChecksum(),
+        expectedChecksumInfo2.getContainerMerkleTree().getDataChecksum());
+
+    // Checksum file exists after container close and matches the expected container
+    // merkle tree for all the datanodes
+    for (HddsDatanodeService hddsDatanode : hddsDatanodes) {
+      GenericTestUtils.waitFor(() -> waitForContainerChecksumToExist(hddsDatanode, containerInfo2), 100, 5000);
+      container2 = (KeyValueContainer) ozoneContainer.getController().getContainer(containerInfo2.getContainerID());
+      ContainerProtos.ContainerChecksumInfo containerChecksumInfo = ContainerMerkleTreeTestUtils.readChecksumFile(
+          container2.getContainerData());
+      ContainerMerkleTreeTestUtils.assertTreesSortedAndMatch(expectedChecksumInfo2.getContainerMerkleTree(),
+          containerChecksumInfo.getContainerMerkleTree());
+    }
+  }
+
+  private boolean waitForContainerChecksumToExist(HddsDatanodeService hddsDatanode,
```

Comment:
There is no wait in this method currently. Probably better to have this wait for container close and then check that the file exists.